src/oversampling_methods/borderline_smote1/borderline_smote1.jl

```diff
@@ -0,0 +1,239 @@
+
+"""
+This function is only called when N>1 and checks whether 0<m<N or not. If m<0, it throws an error.
+and if m>=N, it warns the user and sets m=N-1.
+
+# Arguments
+- `k`: Number of nearest neighbors to consider
+- `n`: Number of observations in the data
+
+# Returns
+-  Number of nearest neighbors to consider
+"""
+function check_m(m, N)
+    if m < 1
+        throw(ERR_NONPOS_M(m))
+    end
+    if m >= N
+        @warn WRN_M_TOO_BIG(m, N)
+        m = N - 1
+    end
+    return m
+end
+
+"""
+Compute a boolean filter according to filter (X, y) points that satisfy the BorderlineSMOTE1 condition.
+
+# Arguments
+- X: A matrix where each row is treated as an observation
+- y: A vector of labels corresponding to the observations
+- m: The number of neighbors to consider while checking the BorderlineSMOTE1 condition. In this, a point may participate
+    in oversampling iff the number of neighbors that belong to its class is in \$(0, m/2]\$
+
+# Returns
+- A boolean filter that can be used to filter the data to remove the points violating the BorderlineSMOTE1 condition.
+"""
+function borderline1_filter(X, y; m=5)
+    m = check_m(m, length(y))
+    tree = BallTree(X)
+    knn_map, _ = knn(tree, X, m + 1, true)
+    knn_matrix = hcat(knn_map...)[2:end, :]
+    # m * num_points mapping from point index to neighbor labels
+    knn_matrix_labels = y[knn_matrix]
+    bool_filter = ones(Bool, length(y))
+    for i in eachindex(y)
+        num_friendly_neighbors = sum(knn_matrix_labels[:, i] .== y[i])
+        # borderline_smote1 condition
+        bool_filter[i] = 0 < num_friendly_neighbors  <= m/2
+    end
+    return BitVector(bool_filter)
+end
+
+
+"""
+    borderline_smote1(
+        X, y;
+        m=5, k=5, ratios=nothing, rng=default_rng(),
+        try_perserve_type=true
+    )
+
+# Description
+Oversamples a dataset using borderline SMOTE1 algorithm to 
+    correct for class imbalance as presented in [1]
+
+# Positional Arguments
+
+$(COMMON_DOCS["INPUTS"])
+
+# Keyword Arguments
+
+- `m::Integer=5`: The number of neighbors to consider while checking the BorderlineSMOTE1 condition. In this, a point may participate
+    in oversampling iff the number of neighbors that belong to its class is in \$(0, m/2]\$. Should be within the range 
```

Comment:
probably "iff" is too geeky for some

```suggestion
    in oversampling if and only if the number of neighbors that belong to its class is in \$(0, m/2]\$. Should be within the range 
```

Reply:
Definitely agree that it's clearer this way. It seems; however, that this has an official entry in the dictionary.